### ISSUE TITLE
Bug: Read-Repair on Cursor/Get only repairs if it hits the node with the most up-to-date data

### ISSUE THREAD

**Issue by etiennedi** [weaviate/weaviate#3762]
tl;dr: The expectation is that paging through all objects with `consistency_level=ALL` should lead to all objects being repaired. However, it seems this is not the case if you send the cursor to the node that is missing data. It only works correctly if you target the node that has the perfect, most up-to-date representation of the data.

## How to reproduce

1. Start a 3-node cluster (ideally with `REPLICATION_MINIMUM_FACTOR=3`)
2. Start a long-running import (e.g. 100k-1M) (set `replication_factor=3` in case you didn't set the global minimum)
3. Kill node 3 during the import and after some time have it rejoin the cluster
4. Wait for the import to finish. Node 3 has missed some insert and now requires repairing. To verify you can either page through all objects on node 3 with `consistency_level=ONE` or you can look at the `/v1/nodes` API which should show a different object count for node 3
5. Now page through all objects with `consistency_level=ALL`. Make sure that this request hits the third node (i.e. the one that's out of sync)

## Expected Behavior
All nodes are back in sync, a read repair has happened on all missing objects. Node 3 should now have the same view of the data as all other nodes.

## Actual Behavior
Node 3 is still missing many objects.

## Additional Notes
* If you would target node 1 or 2 in Step 5 the read-repair works correctly. However, in a real-life setup the user cannot target a specific node. In addition, there might not be a node that has all the correct objects. It could be that some objects are missing on node 1, others missing on node 2, etc. Therefore the expectation is that - no matter which node receives the requests - paging through all objects with `consistency_level=ALL` will always fix all nodes.
* [This script](https://gist.github.com/etiennedi/89e8086f240ee273a1a7d1786c7a8ed2) that runs cursors with various consistency levels might be helpful during reproduction.

**Comment by avirlrma** [comment]
I found out that I'm saying the same thing as @redouan-rhazouani here: https://github.com/weaviate/weaviate/issues/3267#issuecomment-1641831744.

**Comment by etiennedi** [comment]
Thank you for the analysis. The fact that you come to a similar conclusion as @redouan-rhazouani is a good sign :-) 

Would you be interested in contributing the "fix"/feature? If so, we could check if we could arrange a kick-off together with some Weaviate folks that have context on the matter.

**Comment by avirlrma** [comment]
oh definitely @etiennedi. Some guidance on this would be really helpful!

**Comment by avirlrma** [comment]
Hi @etiennedi 
I am preparing a design/solution doc for the implementation and one of the approaches that I got was to use Merkle Trees to find inconsistencies. I would like to know if that’s what the weaviate team had in mind for this.

It will be great if I can get a channel with stakeholders for this, so that I can share my thoughts/design and get team’s opinion on this.

**Comment by avirlrma** [comment]
Dropping a gentle reminder for this.

**Comment by avirlrma** [comment]
While studying more on this, I am realising that hinted handoff might be better for us here. The Merkle tree is complicated to build and represent in weaviate, with that approach we would still require quite a lot of network exchanges to establish inconsistency. 

Hinted handoff can also be thought of as delayed(/async) replication which is more of a natural extension to what we already have in weaviate.

**Comment by etiennedi** [comment]
Please also see https://github.com/weaviate/weaviate/pull/4245 where async replication (using Merkle Trees) is already in progress and scheduled to be released in Weaviate `v1.25` around late March / early April this year.

**Comment by dirkkul** [comment]
PR for async replication is merged, closing the issue. Please reopen in case it is not resolved

**Comment by etiennedi** [comment]
Wondering if we should reopen this. I'm on a case right now where async replication is not enabled (it was created before it existed). So I need to repair using read repairs. It requires 3x the effort (with rf=3) right now because of this bug.

**Comment by dirkkul** [comment]
then let's open it again
